File: leetcode/409. Longest Palindrome/409.cpp
// https://helloacm.com/the-length-of-the-longest-palindrome/
// https://leetcode.com/problems/longest-palindrome/

class Solution {
public:
    int longestPalindrome(string s) {
        unordered_map<char, int> count;
        for (const auto &ch: s) {
            if (count.find(ch) == count.end()) {
                count[ch] = 1;
            } else {
                count[ch] ++;
            }
        }
        int odd = 0, sum = 0;
        for (auto it = count.begin(); it != count.end(); it ++) {
            if (it->second % 2 == 0) {
                sum += it->second;
            } else {
                odd = 1;
                sum += it->second - 1;
            }
        }
        return sum + odd;
    }
};

File: leetcode/409. Longest Palindrome/409.py
# https://helloacm.com/teaching-kids-programming-longest-palindrome-string-from-set-of-characters/
# https://leetcode.com/problems/longest-palindrome/
# EASY, HASH MAP

class Solution:
    def longestPalindrome(self, s: str) -> int:
        d = collections.Counter(s)
        ans = 0
        odd = 0
        for key in d:
            if d[key] % 2 == 0:
                ans += d[key]
            else:
                ans += d[key] - 1
                odd = 1
        ans += odd
        return ans
